File: php/php-wavebox/inc/model/Artist.php
<?php

class Artist extends AbstractItem
{
	// WaveBox Artist fields
	public $artistId;
	public $artistName;
	public $musicBrainzId;

	// API fields
	public $albums = array();
	public $songs = array();
	public $counts = array();

	// Run constructor and create appropriate API client
	public function __construct(AbstractApiClient $apiClient, $clientType = "ArtistsApiClient")
	{
		parent::__construct($apiClient, $clientType);
	}

	// Shortcut to return or lazy-load all albums belonging to this artist
	public function fetchAlbums()
	{
		// Return albums if available
		if (!empty($this->albums))
		{
			return $this->albums;
		}

		// Fetch list of albums belonging to this artist
		$this->albums = $this->apiClient->getArtist($this->artistId, true)->albums;
		return $this->albums;
	}

	// Shortcut to return or lazy-load all songs belonging to this artist
	public function fetchSongs()
	{
		// Return songs if available
		if (!empty($this->songs))
		{
			return $this->songs;
		}

		// Fetch list of songs belonging to this artist
		$this->songs = $this->apiClient->getArtist($this->artistId, false, true)->songs;
		return $this->songs;
	}

	// Shortcut to return or lazy-load all counts belonging to this artist
	public function fetchCounts()
	{
		// Return songs if available
		if (!empty($this->counts))
		{
			return $this->counts;
		}

		// Fetch list of counts belonging to this artist
		$this->counts = $this->apiClient->getAlbumArtist($this->albumArtistId, false, false, true)->counts;
		return $this->counts;
	}
}

File: ruby/rb-wavebox/model/Artist.rb
class Artist
	attr_accessor :item_type_id, :artist_id, :artist_name, :musicbrainz_id

	def initialize(api, artist, albums = nil, songs = nil)
		# API client
		@api = api

		# Artist fields
		@item_type_id = artist['itemTypeId'].to_i()
		@artist_id = artist['artistId'].to_i()
		@artist_name = artist['artistName']
		@musicbrainz_id = artist['musicBrainzId']

		# Album fields
		unless albums.nil?
			@albums = albums.map { |a| Album.new(@api, a) }
		end

		# Song fields
		unless songs.nil?
			@songs = songs.map { |s| Song.new(s) }
		end
	end

	# Fetch all albums belonging to this artist
	def fetch_albums()
		# Return list of albums if already stored
		unless @albums.empty?
			return @albums
		end

		# Fetch and retrieve list of albums
		@albums = @api.fetch_json("artists/" + @artist_id.to_s())['albums'].map { |a| Album.new(@api, a) }
	end

	# Fetch all songs belonging to this artist
	def fetch_songs()
		# Return list of songs if already stored
		unless @songs.empty?
			return @songs
		end

		# Fetch and retrieve list of songs
		@songs = @api.fetch_json(sprintf("artists/%d?includeSongs=true", @artist_id))['songs'].map { |s| Song.new(s) }
	end
end
